tests/mock_vws/test_query.py::TestMaximumImageFileSize::test_png[Real Vuforia] FAILED

See https://github.com/VWS-Python/vws-python-mock/runs/1628674699?check_suite_focus=true

```
        with pytest.raises(requests.exceptions.ConnectionError):
82
>           query(
83
                vuforia_database=vuforia_database,
84
                body=body,
85
            )
86
E           Failed: DID NOT RAISE <class 'requests.exceptions.ConnectionError'>
```

--- a/src/mock_vws/_query_validators/exceptions.py
+++ b/src/mock_vws/_query_validators/exceptions.py
@@ -621,6 +621,33 @@ class ContentLengthHeaderNotInt(ValidatorException):
         }
 
 
+class RequestEntityTooLarge(ValidatorException):
+    """
+    Exception raised when the given image file size is too large.
+    """
+
+    def __init__(self) -> None:
+        """
+        Attributes:
+            status_code: The status code to use in a response if this is
+                raised.
+            response_text: The response text to use in a response if this is
+                raised.
+        """
+        super().__init__()
+        self.status_code = HTTPStatus.REQUEST_ENTITY_TOO_LARGE
+        # TODO assert this!
+        self.response_text = ''
+        date = email.utils.formatdate(None, localtime=False, usegmt=True)
+        self.headers = {
+            'Connection': 'Close',
+            'Date': date,
+            'Server': 'nginx',
+            'Content-Type': 'text/html',
+            'Content-Length': str(len(self.response_text)),
+        }
+
+
 class MatchProcessing(ValidatorException):
     """
     Exception raised a target is matched which is processing or recently

--- a/src/mock_vws/_query_validators/exceptions.py
+++ b/src/mock_vws/_query_validators/exceptions.py
@@ -636,8 +636,17 @@ class RequestEntityTooLarge(ValidatorException):
         """
         super().__init__()
         self.status_code = HTTPStatus.REQUEST_ENTITY_TOO_LARGE
-        # TODO assert this!
-        self.response_text = ''
+        self.response_text = textwrap.dedent(
+            """\
+            <html>\r
+            <head><title>413 Request Entity Too Large</title></head>\r
+            <body bgcolor="white">\r
+            <center><h1>413 Request Entity Too Large</h1></center>\r
+            <hr><center>nginx</center>\r
+            </body>\r
+            </html>\r
+            """,
+        )
         date = email.utils.formatdate(None, localtime=False, usegmt=True)
         self.headers = {
             'Connection': 'Close',

--- a/src/mock_vws/_query_validators/image_validators.py
+++ b/src/mock_vws/_query_validators/image_validators.py
@@ -9,7 +9,7 @@ from typing import Dict
 import requests
 from PIL import Image
 
-from mock_vws._query_validators.exceptions import BadImage, ImageNotGiven
+from mock_vws._query_validators.exceptions import BadImage, ImageNotGiven, RequestEntityTooLarge
 
 
 def validate_image_field_given(
@@ -54,7 +54,7 @@ def validate_image_file_size(
         request_body: The body of the request.
 
     Raises:
-        requests.exceptions.ConnectionError: The image file size is too large.
+        RequestEntityTooLarge: The image file size is too large.
     """
     body_file = io.BytesIO(request_body)
 
@@ -74,7 +74,7 @@ def validate_image_file_size(
     # files.
     max_bytes = 2 * 1024 * 1024
     if len(image) > max_bytes:
-        raise requests.exceptions.ConnectionError
+        raise RequestEntityTooLarge
 
 
 def validate_image_dimensions(

--- a/src/mock_vws/_query_validators/image_validators.py
+++ b/src/mock_vws/_query_validators/image_validators.py
@@ -6,10 +6,13 @@ import cgi
 import io
 from typing import Dict
 
-import requests
 from PIL import Image
 
-from mock_vws._query_validators.exceptions import BadImage, ImageNotGiven, RequestEntityTooLarge
+from mock_vws._query_validators.exceptions import (
+    BadImage,
+    ImageNotGiven,
+    RequestEntityTooLarge,
+)
 
 
 def validate_image_field_given(

--- a/tests/mock_vws/test_authorization_header.py
+++ b/tests/mock_vws/test_authorization_header.py
@@ -62,6 +62,7 @@ class TestAuthorizationHeader:
                 content_type='text/plain;charset=iso-8859-1',
                 cache_control=None,
                 www_authenticate='VWS',
+                connection='keep-alive',
             )
             assert response.text == 'Authorization header missing.'
             return
@@ -114,6 +115,7 @@ class TestMalformed:
                 content_type='text/plain;charset=iso-8859-1',
                 cache_control=None,
                 www_authenticate='VWS',
+                connection='keep-alive',
             )
             assert response.text == 'Malformed authorization header.'
             return
@@ -161,6 +163,7 @@ class TestMalformed:
                 content_type='text/html;charset=iso-8859-1',
                 cache_control='must-revalidate,no-cache,no-store',
                 www_authenticate=None,
+                connection='keep-alive',
             )
             content_filename = 'jetty_error_array_out_of_bounds.html'
             content_filename_2 = 'jetty_error_array_out_of_bounds_2.html'
@@ -227,6 +230,7 @@ class TestBadKey:
             content_type='application/json',
             cache_control=None,
             www_authenticate='VWS',
+            connection='keep-alive',
         )
 
         assert response.json().keys() == {'transaction_id', 'result_code'}
@@ -285,6 +289,7 @@ class TestBadKey:
             content_type='application/json',
             cache_control=None,
             www_authenticate='VWS',
+            connection='keep-alive',
         )
 
         assert response.json().keys() == {'transaction_id', 'result_code'}

--- a/tests/mock_vws/test_content_length.py
+++ b/tests/mock_vws/test_content_length.py
@@ -97,6 +97,7 @@ class TestIncorrect:
                 content_type='application/json',
                 cache_control=None,
                 www_authenticate='VWS',
+                connection='keep-alive',
             )
             return
 

--- a/tests/mock_vws/test_date_header.py
+++ b/tests/mock_vws/test_date_header.py
@@ -75,6 +75,7 @@ class TestMissing:
                 content_type=expected_content_type,
                 cache_control=None,
                 www_authenticate=None,
+                connection='keep-alive',
             )
             return
 
@@ -141,6 +142,7 @@ class TestFormat:
                 content_type='text/plain;charset=iso-8859-1',
                 cache_control=None,
                 www_authenticate='VWS',
+                connection='keep-alive',
             )
             return
 

--- a/tests/mock_vws/test_invalid_json.py
+++ b/tests/mock_vws/test_invalid_json.py
@@ -99,6 +99,7 @@ class TestInvalidJSON:
                 content_type='application/json',
                 cache_control=None,
                 www_authenticate=None,
+                connection='keep-alive',
             )
             expected_text = 'No image.'
             assert response.text == expected_text

--- a/tests/mock_vws/test_query.py
+++ b/tests/mock_vws/test_query.py
@@ -228,6 +228,7 @@ class TestContentType:
             content_type=resp_content_type,
             cache_control=resp_cache_control,
             www_authenticate=None,
+            connection='keep-alive',
         )
 
     def test_incorrect_with_boundary(
@@ -284,6 +285,7 @@ class TestContentType:
             content_type=None,
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
 
     @pytest.mark.parametrize(
@@ -347,6 +349,7 @@ class TestContentType:
             content_type='text/html;charset=utf-8',
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
 
     def test_bogus_boundary(
@@ -398,6 +401,7 @@ class TestContentType:
             content_type='application/json',
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
 
     def test_extra_section(
@@ -590,6 +594,7 @@ class TestIncorrectFields:
             content_type='application/json',
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
 
     def test_extra_fields(
@@ -615,6 +620,7 @@ class TestIncorrectFields:
             status_code=HTTPStatus.BAD_REQUEST,
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
 
     def test_missing_image_and_extra_fields(
@@ -640,6 +646,7 @@ class TestIncorrectFields:
             status_code=HTTPStatus.BAD_REQUEST,
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
 
 
@@ -778,6 +785,7 @@ class TestMaxNumResults:
             status_code=HTTPStatus.BAD_REQUEST,
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
 
     @pytest.mark.parametrize(
@@ -816,6 +824,7 @@ class TestMaxNumResults:
             status_code=HTTPStatus.BAD_REQUEST,
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
 
 
@@ -999,6 +1008,7 @@ class TestIncludeTargetData:
             content_type='application/json',
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
 
 
@@ -1111,6 +1121,7 @@ class TestAcceptHeader:
             content_type=None,
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
 
 
@@ -1189,6 +1200,7 @@ class TestBadImage:
             content_type='application/json',
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
         assert response.json().keys() == {'transaction_id', 'result_code'}
         assert_valid_transaction_id(response=response)
@@ -1221,7 +1233,7 @@ class TestMaximumImageFileSize:
         https://library.vuforia.com/articles/Solution/How-To-Perform-an-Image-Recognition-Query.
         the maximum file size is "2MiB for PNG".
 
-        Above this limit, a ``TODO`` is raised.
+        Above this limit, a ``REQUEST_ENTITY_TOO_LARGE`` response is returned.
         We do not test exactly at this limit, but that may be beneficial in the
         future.
         """
@@ -1279,9 +1291,10 @@ class TestMaximumImageFileSize:
         assert_vwq_failure(
             response=response,
             status_code=HTTPStatus.REQUEST_ENTITY_TOO_LARGE,
-            content_type=None,
+            content_type='text/html',
             cache_control=None,
             www_authenticate=None,
+            connection='Close',
         )
 
     def test_jpeg(
@@ -1343,11 +1356,19 @@ class TestMaximumImageFileSize:
         assert image_content_size > max_bytes
         assert (image_content_size * 0.95) < max_bytes
 
-        with pytest.raises(requests.exceptions.ConnectionError):
-            query(
-                vuforia_database=vuforia_database,
-                body=body,
-            )
+        response = query(
+            vuforia_database=vuforia_database,
+            body=body,
+        )
+
+        assert_vwq_failure(
+            response=response,
+            status_code=HTTPStatus.REQUEST_ENTITY_TOO_LARGE,
+            content_type='text/html',
+            cache_control=None,
+            www_authenticate=None,
+            connection='Close',
+        )
 
 
 @pytest.mark.usefixtures('verify_mock_vuforia')
@@ -1398,6 +1419,7 @@ class TestMaximumImageDimensions:
             content_type='application/json',
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
         assert response.json().keys() == {'transaction_id', 'result_code'}
         assert_valid_transaction_id(response=response)
@@ -1456,6 +1478,7 @@ class TestMaximumImageDimensions:
             content_type='application/json',
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
         assert response.json().keys() == {'transaction_id', 'result_code'}
         assert_valid_transaction_id(response=response)
@@ -1525,6 +1548,7 @@ class TestImageFormats:
             content_type='application/json',
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
         assert response.json().keys() == {'transaction_id', 'result_code'}
         assert_valid_transaction_id(response=response)
@@ -1723,6 +1747,7 @@ class TestDeleted:
                     status_code=HTTPStatus.INTERNAL_SERVER_ERROR,
                     cache_control='must-revalidate,no-cache,no-store',
                     www_authenticate=None,
+                    connection='keep-alive',
                 )
 
                 return
@@ -1962,6 +1987,7 @@ class TestInactiveProject:
             content_type='application/json',
             cache_control=None,
             www_authenticate=None,
+            connection='keep-alive',
         )
         assert response.json().keys() == {'transaction_id', 'result_code'}
         assert_valid_transaction_id(response=response)

--- a/tests/mock_vws/test_query.py
+++ b/tests/mock_vws/test_query.py
@@ -62,6 +62,17 @@ _JETTY_CONTENT_TYPE_ERROR = textwrap.dedent(
     """,  # noqa: E501
 )
 
+_NGINX_REQUEST_ENTITY_TOO_LARGE_ERROR = textwrap.dedent(
+    """\
+    <html>\r
+    <head><title>413 Request Entity Too Large</title></head>\r
+    <body bgcolor="white">\r
+    <center><h1>413 Request Entity Too Large</h1></center>\r
+    <hr><center>nginx</center>\r
+    </body>\r
+    </html>\r
+    """,
+)
 
 _JETTY_ERROR_DELETION_NOT_COMPLETE_START_PATH = (
     Path(__file__).parent / 'jetty_error_deletion_not_complete.html'
@@ -1296,6 +1307,7 @@ class TestMaximumImageFileSize:
             www_authenticate=None,
             connection='Close',
         )
+        assert response.text == _NGINX_REQUEST_ENTITY_TOO_LARGE_ERROR
 
     def test_jpeg(
         self,
@@ -1369,6 +1381,8 @@ class TestMaximumImageFileSize:
             www_authenticate=None,
             connection='Close',
         )
+
+        assert response.text == _NGINX_REQUEST_ENTITY_TOO_LARGE_ERROR
 
 
 @pytest.mark.usefixtures('verify_mock_vuforia')

--- a/tests/mock_vws/test_query.py
+++ b/tests/mock_vws/test_query.py
@@ -1221,7 +1221,7 @@ class TestMaximumImageFileSize:
         https://library.vuforia.com/articles/Solution/How-To-Perform-an-Image-Recognition-Query.
         the maximum file size is "2MiB for PNG".
 
-        Above this limit, a ``ConnectionError`` is raised.
+        Above this limit, a ``TODO`` is raised.
         We do not test exactly at this limit, but that may be beneficial in the
         future.
         """
@@ -1271,11 +1271,18 @@ class TestMaximumImageFileSize:
         assert image_content_size > max_bytes
         assert (image_content_size * 0.95) < max_bytes
 
-        with pytest.raises(requests.exceptions.ConnectionError):
-            query(
-                vuforia_database=vuforia_database,
-                body=body,
-            )
+        response = query(
+            vuforia_database=vuforia_database,
+            body=body,
+        )
+
+        assert_vwq_failure(
+            response=response,
+            status_code=HTTPStatus.REQUEST_ENTITY_TOO_LARGE,
+            content_type=None,
+            cache_control=None,
+            www_authenticate=None,
+        )
 
     def test_jpeg(
         self,

--- a/tests/mock_vws/test_query.py
+++ b/tests/mock_vws/test_query.py
@@ -1307,7 +1307,7 @@ class TestMaximumImageFileSize:
         the maximum file size is "512 KiB for JPEG".
         However, this test shows that the maximum size for JPEG is 2 MiB.
 
-        Above this limit, a ``ConnectionError`` is raised.
+        Above this limit, a ``REQUEST_ENTITY_TOO_LARGE`` response is returned.
         We do not test exactly at this limit, but that may be beneficial in the
         future.
         """

--- a/tests/mock_vws/test_unexpected_json.py
+++ b/tests/mock_vws/test_unexpected_json.py
@@ -77,6 +77,7 @@ class TestUnexpectedJSON:
                 content_type=None,
                 cache_control=None,
                 www_authenticate=None,
+                connection='keep-alive',
             )
             return
 

--- a/tests/mock_vws/utils/assertions.py
+++ b/tests/mock_vws/utils/assertions.py
@@ -202,6 +202,7 @@ def assert_vwq_failure(
     content_type: str | None,
     cache_control: str | None,
     www_authenticate: str | None,
+    connection: str,
 ) -> None:
     """
     Assert that a VWQ failure response is as expected.
@@ -212,6 +213,7 @@ def assert_vwq_failure(
         status_code: The expected status code.
         cache_control: The expected Cache-Control header.
         www_authenticate: The expected WWW-Authenticate header.
+        connection: The expected Connection header.
 
     Raises:
         AssertionError: The response is not in the expected VWQ error format
@@ -248,7 +250,7 @@ def assert_vwq_failure(
         response_header_keys_chunked,
     )
     assert response.headers.get('transfer-encoding', 'chunked') == 'chunked'
-    assert response.headers['Connection'] == 'keep-alive'
+    assert response.headers['Connection'] == connection
     if 'Content-Length' in response.headers:
         assert response.headers['Content-Length'] == str(len(response.text))
     assert_valid_date_header(response=response)

